### PR TITLE
TF: GPT2 with native embedding layers

### PR DESCRIPTION
**Pull request by gante**
# What does this PR do?

This PR continues the (paused) goal of deprecating our custom TF embedding layers and related code. Previously, we have converted encoder-decoder models (e.g. [here](https://github.com/huggingface/transformers/pull/19263)), removing `TFSharedEmbeddings` there and making the necessary adaptations.

In this PR, I make the necessary adaptations for GPT2. The goal is for you, the reviewers, to raise objections in this PR :D All slow tests for TF GPT2 are passing.

Then, the following sequence of PRs will be opened:
1. Remove `TFSharedEmbeddings` from the other decoder-only models
2. Remove other uses of `TFSharedEmbeddings` in the codebase (e.g. in tests)
3. Remove `resize_token_embeddings` and all related functions (it is only used to resize our models' embeddings instantiated with `TFSharedEmbeddings`)
4. Remove the slow decorator from `test_save_load_after_resize_token_embeddings`, which will be fixed as a consequence of these changes 🙌

### CODE REVIEW

File: src/transformers/modeling_tf_utils.py [hunk 1]
```diff
@@ -3132,6 +3132,10 @@ def __init__(self, vocab_size: int, hidden_size: int, initializer_range: Optiona
         self.vocab_size = vocab_size
         self.hidden_size = hidden_size
         self.initializer_range = hidden_size**-0.5 if initializer_range is None else initializer_range
+        warnings.warn(
+            "TFSharedEmbeddings is scheduled for deletion, use `tf.keras.layers.Embedding` instead.",
```
Comment (gante) [inline src/transformers/modeling_tf_utils.py:3136]:
```suggestion
            "TFSharedEmbeddings are scheduled for deletion, use `tf.keras.layers.Embedding` instead.",
```

File: src/transformers/models/gpt2/modeling_tf_gpt2.py [hunk 2]
```diff
@@ -1048,7 +1045,8 @@ def call(
             all_hidden_states = transformer_outputs.hidden_states[:-1] + (hidden_states,)
         else:
             all_hidden_states = None
-        lm_logits = self.transformer.wte(hidden_states, mode="linear")
+
```
Comment (gante) [inline src/transformers/models/gpt2/modeling_tf_gpt2.py:1048]:
```suggestion
```

File: src/transformers/modeling_tf_utils.py [hunk 3]
```diff
@@ -3132,6 +3132,10 @@ def __init__(self, vocab_size: int, hidden_size: int, initializer_range: Optiona
         self.vocab_size = vocab_size
         self.hidden_size = hidden_size
         self.initializer_range = hidden_size**-0.5 if initializer_range is None else initializer_range
+        warnings.warn(
+            "TFSharedEmbeddings are scheduled for deletion, use `tf.keras.layers.Embedding` instead.",
```
Comment (amyeroberts) [inline src/transformers/modeling_tf_utils.py:3136]: Could you put in a specific version in the warning? Typically >= 2 versions from the release this will be in

**Review by amyeroberts — APPROVED** [review]
Nice - thanks for adding!

**Review by Rocketknight1 — APPROVED** [review]
Looks clean to me!

**Comment by arivero** [comment]
I noticed this change one year later, while doing a demo of LoRA in the embedding layer :-D  Nice, but I am always afraid of doing tensorflow operations outside of keras layers. Usually masks are lost, as well as any subtle things that depend of chaining layers. It is not a problem in transformers because the -100 trick is the standard mask and the loss is aware of it.
